### PR TITLE
[GCP FileShare Support] Implement Fileshare CRUD API Support for GCP Fileshare Driver

### LINKED ISSUE #1065: Fileshare support for GCP
**Issue/Feature Description:**
Add another backend GCP for Fileshare support

**Why this issue to fixed / feature is needed(give scenarios or use cases):**
Add support for basic CRUD operations

**How to reproduce, in case of a bug:**

**Other Notes /  Environment Information: (Please give the env information, log link or any useful information for this issue)**

### PR DESCRIPTION
**Pull request by himanshuvar**
**What type of PR is this?**
> /kind new feature

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**: This PR is for GCP FileShare Support in multi-cloud. This PR is having the following highlights

1. Enhance Create FileShare API for GCP backend
2. Implement List FileShare API for GCP backend
3. Implement Get FileShare API for GCP backend
4. Implement Put FileShare API for GCP backend
5. Implement Delete FileShare API for GCP backend
6. Add Wait Poll mechanishm for GCP Filestore operation status completion check

**Which issue(s) this PR fixes**:

Fixes #1065 #1077 #1078 #1079 #1080

**Test Report Added?**:
> /kind TESTED

**Test Report**:

The following testing is done for Create File Share API for GCP cloud backend:

- [x] Register Backend

- [x] List Backend

- [x] Create FileShare for GCP Backend

- [x] List FileShares for GCP Backend

- [x] Get FileShare for GCP Backend

- [x] Update FileShare for GCP Backend

- [x]  Delete FileShare for GCP Backend

**Register a Backend**

Registering an GCP Backend for File Share

POST http://192.168.20.162:8089/v1/94b280022d0c4401bcf3b0ea85870519/backends

Request:
```
{
  "name": "gcp-backend-file",
  "type": "gcp-file",
  "Region": "<project_id>",
  "Access": "<access_key>",
  "Security": "<secret_key>"
}
```
Response:
```
{
    "id": "5f67f7c9f969b6000157cdde",
    "tenantId": "94b280022d0c4401bcf3b0ea85870519",
    "userId": "558057c4256545bd8a307c37464003c9",
    "name": "gcp-backend-file",
    "type": "gcp-file",
    "region": "<project_id>",
    "access": "<access_key>",
    "security": "<secret_key>"
}
```
**List Backend:**

GET http://192.168.20.162:8089/v1/94b280022d0c4401bcf3b0ea85870519/backends

Response:

```
{
    "backends": [
        {
            "id": "5f67f7c9f969b6000157cdde",
            "tenantId": "94b280022d0c4401bcf3b0ea85870519",
            "userId": "558057c4256545bd8a307c37464003c9",
            "name": "gcp-backend-file",
            "type": "gcp-file",
            "region": "<project_id>"
        }
    ],
    "next": 1
}
```
**Create a File share**

Creating File Share for GCP Backend

POST http://192.168.20.162:8089/v1/94b280022d0c4401bcf3b0ea85870519/file/shares

Request:

```
{
    "name": "gcpfshim",
    "description": "GCP FileShare",
    "backendId": "5f67f7c9f969b6000157cdde",
    "availabilityZone": "asia-east1-a",
    "size": 1099511627776,
    "type": "cloudFS",
    "tags": [
      {
        "key": "creator",
        "value": "himanshu"
      }
    ],
    "metadata": {
      "Tier": "STANDARD"
    }
}
```

Response:

```
{
    "id": "5f67f7fbf8d41900017cb022",
    "createdAt": "2020-09-21T06:16:51",
    "updatedAt": "2020-09-21T06:16:51",
    "name": "gcpfshim",
    "description": "GCP FileShare",
    "tenantId": "94b280022d0c4401bcf3b0ea85870519",
    "userId": "558057c4256545bd8a307c37464003c9",
    "backendId": "5f67f7c9f969b6000157cdde",
    "backend": "gcp-backend-file",
    "size": 1099511627776,
    "type": "cloudFS",
    "region": "<project_id>",
    "availabilityZone": "asia-east1-a",
    "status": "creating",
    "tags": [
        {
            "key": "creator",
            "value": "himanshu"
        }
    ],
    "metadata": {
        "fields": {
            "Tier": {
                "Kind": {
                    "StringValue": "STANDARD"
                }
            }
        }
    }
}
```
image
image

**List File shares**

GET http://192.168.20.162:8089/v1/94b280022d0c4401bcf3b0ea85870519/file/shares

Response:

```
{
    "fileshares": [
        {
            "id": "5f67f7fbf8d41900017cb022",
            "createdAt": "2020-09-21T06:16:51",
            "updatedAt": "2020-09-21T06:20:02",
            "name": "gcpfshim",
            "description": "GCP FileShare",
            "tenantId": "94b280022d0c4401bcf3b0ea85870519",
            "userId": "558057c4256545bd8a307c37464003c9",
            "backendId": "5f67f7c9f969b6000157cdde",
            "backend": "gcp-backend-file",
            "size": 1099511627776,
            "type": "cloudFS",
            "region": "<project_id>",
            "availabilityZone": "asia-east1-a",
            "status": "available",
            "tags": [
                {
                    "key": "creator",
                    "value": "himanshu"
                }
            ],
            "metadata": {
                "fields": {
                    "CreationTimeAtBackend": {
                        "Kind": {
                            "StringValue": "2020-09-21T00:46:51.121497622Z"
                        }
                    },
                    "Etag": {
                        "Kind": {
                            "StringValue": ""
                        }
                    },
                    "InstanceResourceName": {
                        "Kind": {
                            "StringValue": "<resourceName>"
                        }
                    },
                    "Networks": {
                        "Kind": {
                            "StructValue": {
                                "fields": {
                                    "network": {
                                        "Kind": {
                                            "StringValue": "default"
                                        }
                                    },
                                    "reservedIpRange": {
                                        "Kind": {
                                            "StringValue": "<IP>"
                                        }
                                    }
                                }
                            }
                        }
                    },
                    "State": {
                        "Kind": {
                            "StringValue": "READY"
                        }
                    },
                    "StatusMessage": {
                        "Kind": {
                            "StringValue": ""
                        }
                    },
                    "Tier": {
                        "Kind": {
                            "StringValue": "STANDARD"
                        }
                    },
                    "ipAddresses": {
                        "Kind": {
                            "StringValue": "<IPAddress>"
                        }
                    },
                    "modes": {
                        "Kind": {
                            "StringValue": "MODE_IPV4"
                        }
                    }
                }
            }
        }
    ],
    "next": 1
}
```

**Get File share**

GET http://192.168.20.162:8089/v1/94b280022d0c4401bcf3b0ea85870519/file/shares/5f67f7fbf8d41900017cb022

Response:

```
{
    "id": "5f67f7fbf8d41900017cb022",
    "createdAt": "2020-09-21T06:16:51",
    "updatedAt": "2020-09-21T06:20:02",
    "name": "gcpfshim",
    "description": "GCP FileShare",
    "tenantId": "94b280022d0c4401bcf3b0ea85870519",
    "userId": "558057c4256545bd8a307c37464003c9",
    "backendId": "5f67f7c9f969b6000157cdde",
    "backend": "gcp-backend-file",
    "size": 1099511627776,
    "type": "cloudFS",
    "region": "<project_id>",
    "availabilityZone": "asia-east1-a",
    "status": "available",
    "tags": [
        {
            "key": "creator",
            "value": "himanshu"
        }
    ],
    "metadata": {
        "fields": {
            "CreationTimeAtBackend": {
                "Kind": {
                    "StringValue": "2020-09-21T00:46:51.121497622Z"
                }
            },
            "Etag": {
                "Kind": {
                    "StringValue": ""
                }
            },
            "InstanceResourceName": {
                "Kind": {
                    "StringValue": "<resourceName>"
                }
            },
            "Networks": {
                "Kind": {
                    "StructValue": {
                        "fields": {
                            "network": {
                                "Kind": {
                                    "StringValue": "default"
                                }
                            },
                            "reservedIpRange": {
                                "Kind": {
                                    "StringValue": "<IP>"
                                }
                            }
                        }
                    }
                }
            },
            "State": {
                "Kind": {
                    "StringValue": "READY"
                }
            },
            "StatusMessage": {
                "Kind": {
                    "StringValue": ""
                }
            },
            "Tier": {
                "Kind": {
                    "StringValue": "STANDARD"
                }
            },
            "ipAddresses": {
                "Kind": {
                    "StringValue": "<IPAddress>"
                }
            },
            "modes": {
                "Kind": {
                    "StringValue": "MODE_IPV4"
                }
            }
        }
    }
}
```

**Update a File share**

Updating File Share for GCP Backend

PUT http://192.168.20.162:8089/v1/94b280022d0c4401bcf3b0ea85870519/file/shares/5f67f7fbf8d41900017cb022

Request:

```
{
	"description": "GCP FileShare Updated",
	"size": 2199023255552
}
```
Response:
```
{
    "id": "5f67f7fbf8d41900017cb022",
    "createdAt": "2020-09-21T06:16:51",
    "updatedAt": "2020-09-21T06:26:22",
    "name": "gcpfshim",
    "description": "GCP FileShare Updated",
    "tenantId": "94b280022d0c4401bcf3b0ea85870519",
    "userId": "558057c4256545bd8a307c37464003c9",
    "backendId": "5f67f7c9f969b6000157cdde",
    "backend": "gcp-backend-file",
    "size": 2199023255552,
    "type": "cloudFS",
    "region": "<project_id>",
    "availabilityZone": "asia-east1-a",
    "status": "updating",
    "metadata": {
        "fields": {
            "CreationTimeAtBackend": {
                "Kind": {
                    "StringValue": "2020-09-21T00:46:51.121497622Z"
                }
            },
            "Etag": {
                "Kind": {
                    "StringValue": ""
                }
            },
            "InstanceResourceName": {
                "Kind": {
                    "StringValue": "<resourceName>"
                }
            },
            "Networks": {
                "Kind": {
                    "StructValue": {
                        "fields": {
                            "network": {
                                "Kind": {
                                    "StringValue": "default"
                                }
                            },
                            "reservedIpRange": {
                                "Kind": {
                                    "StringValue": "<IP>"
                                }
                            }
                        }
                    }
                }
            },
            "State": {
                "Kind": {
                    "StringValue": "READY"
                }
            },
            "StatusMessage": {
                "Kind": {
                    "StringValue": ""
                }
            },
            "Tier": {
                "Kind": {
                    "StringValue": "STANDARD"
                }
            },
            "ipAddresses": {
                "Kind": {
                    "StringValue": "<IPAddress>"
                }
            },
            "modes": {
                "Kind": {
                    "StringValue": "MODE_IPV4"
                }
            }
        }
    }
}
```
image

**Delete File share**

DELETE http://192.168.20.162:8089/v1/94b280022d0c4401bcf3b0ea85870519/file/shares/5f67f7fbf8d41900017cb022

Response:

```
{}
```
image
image

**Special notes for your reviewer**:

### CODE REVIEW

File: contrib/datastore/file/gcp/gcp.go [hunk 1]
```diff
@@ -85,27 +94,232 @@ func (g GcpAdapter) CreateFileShare(ctx context.Context, fs *file.CreateFileShar
 }
 
 func (g GcpAdapter) GetFileShare(ctx context.Context, fs *file.GetFileShareRequest) (*file.GetFileShareResponse, error) {
-	panic("implement me")
+	isCreateOpDone, err := g.isOperationDone(createOp)
```
Comment (kumarashit) [inline contrib/datastore/file/gcp/gcp.go:97]: Why do we need to check the completion of all these operations for getting fileshare info? Please put appropriate comment

File: contrib/datastore/file/gcp/gcp.go [hunk 2]
```diff
@@ -85,27 +94,232 @@ func (g GcpAdapter) CreateFileShare(ctx context.Context, fs *file.CreateFileShar
...
+
+	if instance.State == "READY" {
+		// GCP FileStore Instance exists & ready to use
+		fileShare.Status = "available"
+		return &file.GetFileShareResponse{
+			Fileshare: fileShare,
+		}, nil
+	} else {
+		// GCP FileStore Instance exists but is not ready to use
+		return &file.GetFileShareResponse{
```
Comment (kumarashit) [inline contrib/datastore/file/gcp/gcp.go:136]: If the fileshare is not ready, it means provider is already tried to initialize or trying to make it ready, So the status should be not ready. Right? If we send error, what does the API shows as o/p

File: contrib/datastore/file/gcp/gcp.go [hunk 3]
```diff
@@ -85,27 +94,232 @@ func (g GcpAdapter) CreateFileShare(ctx context.Context, fs *file.CreateFileShar
...
+			Fileshare: fileShare,
+		}, fmt.Errorf("instance %s is in %s", instance.Name, instance.State)
+	}
 }
 
 func (g GcpAdapter) ListFileShare(ctx context.Context, fs *file.ListFileShareRequest) (*file.ListFileShareResponse, error) {
-	panic("implement me")
+	instances, err := g.instancesService.List(g.locationURI(g.backend.Region, "-")).Context(ctx).Do()
+	if err != nil {
+		log.Error(err)
```
Comment (kumarashit) [inline contrib/datastore/file/gcp/gcp.go:145]: This will simply log the error sent by GCP, right?

File: contrib/datastore/file/gcp/gcp.go (hunk 1 above)
Comment (himanshuvar) [reply contrib/datastore/file/gcp/gcp.go:97]: Sync DB require to update the resource metadata using GET call from cloud backend. That's why these check are required here.

**Review by kumarashit — APPROVED** [review]
LGTM

File: contrib/datastore/file/gcp/gcp.go (hunk 2 above)
Comment (himanshuvar) [reply contrib/datastore/file/gcp/gcp.go:136]: Could you please elaborate which provider?
Else is taking care of sending the error for any STATE NOT READY for FS instance.

File: contrib/datastore/file/gcp/gcp.go (hunk 3 above)
Comment (himanshuvar) [reply contrib/datastore/file/gcp/gcp.go:145]: Yes, It will log the message in driver before returning to service layer

**Comment by himanshuvar** [comment]
> 
> 
> LGTM

Done

File: contrib/datastore/file/gcp/gcp.go (hunk 2 above)
Comment (kumarashit) [reply contrib/datastore/file/gcp/gcp.go:136]: GCP Cloud provider.

**Review by kumarashit — APPROVED** [review]
LGTM

File: contrib/datastore/file/gcp/gcp.go (hunk 2 above)
Comment (himanshuvar) [reply contrib/datastore/file/gcp/gcp.go:136]: Yes, GCP FS operations are submitted to the cloud provider. Now, polling takes care of checking the status of the FS instance and waits for getting it ready from cloud provider. If there is any failure in the operation, Else clause will take care of returning the Error to the service layer.
If we send error, API update the FS state to "error" & return

**Review by skdwriting — APPROVED** [review]
LGTM

**Pull request merged**